File: backend/app/schemas/analytics.py
from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field


class PatternResponse(BaseModel):
    pattern_type: str
    title: str
    description: str
    severity: str
    data: dict[str, object] = Field(default_factory=dict)
    locked: bool = False


class PatternsEnvelope(BaseModel):
    patterns: list[PatternResponse] = Field(default_factory=list)
    total_completed_trades: int
    threshold: int
    unlocked: bool


class AnalyzePatternsResponse(BaseModel):
    patterns: list[PatternResponse] = Field(default_factory=list)
    total_completed_trades: int
    threshold: int
    unlocked: bool


class TradeExtremes(BaseModel):
    symbol: str | None = None
    pnl: float | None = None
    exit_date: date | None = None


class MonthlyPnlPoint(BaseModel):
    month: str
    pnl: float


class AnalyticsSummaryResponse(BaseModel):
    total_trades: int
    win_rate: float
    total_pnl: float
    avg_pnl_per_trade: float
    best_trade: TradeExtremes
    worst_trade: TradeExtremes
    avg_holding_days: float
    most_traded_symbol: str | None = None
    monthly_pnl: list[MonthlyPnlPoint] = Field(default_factory=list)

File: frontend/src/lib/analytics.ts
import { apiFetch } from "@/lib/api";

export interface PatternResponse {
  pattern_type: string;
  title: string;
  description: string;
  severity: "high" | "medium" | "low";
  data: Record<string, unknown>;
  locked: boolean;
}

export interface PatternsEnvelope {
  patterns: PatternResponse[];
  total_completed_trades: number;
  threshold: number;
  unlocked: boolean;
}

export interface AnalyticsSummaryResponse {
  total_trades: number;
  win_rate: number;
  total_pnl: number;
  avg_pnl_per_trade: number;
  best_trade: {
    symbol: string | null;
    pnl: number | null;
    exit_date: string | null;
  };
  worst_trade: {
    symbol: string | null;
    pnl: number | null;
    exit_date: string | null;
  };
  avg_holding_days: number;
  most_traded_symbol: string | null;
  monthly_pnl: Array<{
    month: string;
    pnl: number;
  }>;
}

export function getAnalyticsSummary(): Promise<AnalyticsSummaryResponse> {
  return apiFetch<AnalyticsSummaryResponse>("/analytics/summary");
}

export function getPatterns(): Promise<PatternsEnvelope> {
  return apiFetch<PatternsEnvelope>("/analytics/patterns");
}
